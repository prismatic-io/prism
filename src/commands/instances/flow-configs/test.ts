import { Command, Flags, Args, ux } from "@oclif/core";
import { gqlRequest, gql } from "../../../graphql";

interface LogNode {
  [index: string]: unknown;
  timestamp: string;
  severity: string;
  message: string;
}

interface FetchLogsResult {
  logs: LogNode[];
  cursor: string | undefined;
  executionComplete: boolean | undefined;
}

export default class TestCommand extends Command {
  static description = "Test a Flow Config of an Instance";
  static args = {
    flowConfig: Args.string({
      description: "ID of a Flow Config to test",
      required: true,
    }),
  };

  static flags = {
    ...ux.table.flags({ only: ["extended", "columns"] }),
    tail: Flags.boolean({
      required: false,
      char: "t",
      description: "Tail logs of the flow config test run",
    }),
    payload: Flags.string({
      required: false,
      char: "p",
      description:
        "Optional JSON-formatted data payload to submit with the test",
    }),
    contentType: Flags.string({
      required: false,
      char: "c",
      description: "Optional content-type for the test payload",
    }),
  };

  async run() {
    const {
      args: { flowConfig },
      flags: { tail, payload, contentType },
    } = await this.parse(TestCommand);

    const result = await gqlRequest({
      document: gql`
        mutation testInstanceFlowConfig(
          $id: ID!
          $payload: String
          $contentType: String
        ) {
          testInstanceFlowConfig(
            input: { id: $id, payload: $payload, contentType: $contentType }
          ) {
            testInstanceFlowConfigResult {
              flowConfig {
                id
              }
              execution {
                id
              }
            }
            errors {
              field
              messages
            }
          }
        }
      `,
      variables: {
        id: flowConfig,
        payload,
        contentType,
      },
    });

    if (!tail) {
      return;
    }

    const executionId =
      result.testInstanceFlowConfig.testInstanceFlowConfigResult.execution.id;
    await this.tailLogs(executionId);
  }

  private async tailLogs(executionId: string) {
    const { flags } = await this.parse(TestCommand);

    let nextCursor: string | undefined = undefined;
    // eslint-disable-next-line no-constant-condition
    while (true) {
      await ux.wait(500);

      const result: any = await this.fetchLogs(executionId, nextCursor);
      if (result === undefined) continue;

      const { logs, cursor, executionComplete } = result;

      nextCursor = cursor;

      ux.table(
        logs,
        {
          timestamp: {},
          severity: {
            minWidth: 12,
          },
          message: {},
        },
        { ...flags, "no-header": true }
      );

      if (executionComplete) return;
    }
  }

  private async fetchLogs(
    executionId: string,
    nextCursor?: string
  ): Promise<FetchLogsResult | undefined> {
    const results = await gqlRequest({
      document: gql`
        query listInstanceTestLogs($executionId: ID!, $nextCursor: String) {
          logs(
            executionResult: $executionId
            after: $nextCursor
            orderBy: { field: TIMESTAMP, direction: ASC }
          ) {
            edges {
              node {
                timestamp
                severity
                message
              }
              cursor
            }
          }
        }
      `,
      variables: {
        executionId,
        nextCursor,
      },
    });

    const {
      edges,
    }: {
      edges: { cursor?: string; node: LogNode }[];
    } = results.logs;
    if (!edges || edges.length === 0) {
      return undefined;
    }

    const logs = edges.map(({ node }) => node);
    const executionComplete = logs.reduce<boolean>(
      (result: boolean, { message }) =>
        result || message.startsWith("Ending Instance Execution"),
      false
    );

    const { cursor } = edges[edges.length - 1];
    return { logs, cursor, executionComplete };
  }
}
