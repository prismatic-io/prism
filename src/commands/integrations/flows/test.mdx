```bash
# Test an integration flow with a payload file and tail the logs and step results
prism integrations:flows:test \
    -p=some_payload_file.xml
    -c=application/xml
    --tail-logs
    --tail-results
```
