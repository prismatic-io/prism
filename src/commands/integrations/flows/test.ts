import { Command, Flags, Args, ux } from "@oclif/core";
import { gql, gqlRequest } from "../../../graphql.js";

interface LogNode {
  [index: string]: unknown;
  timestamp: string;
  severity: string;
  message: string;
}

interface FetchLogsResult {
  logs: LogNode[];
  cursor: string | undefined;
  executionComplete: boolean;
}

export default class TestCommand extends Command {
  static description = "Run a test of an Integration Flow";
  static args = {
    flow: Args.string({ description: "ID of a flow to test", required: true }),
  };

  static flags = {
    ...ux.table.flags({ only: ["extended", "columns"] }),
    tail: Flags.boolean({
      char: "t",
      description: "Tail logs of the integration test run",
      required: false,
    }),
    payload: Flags.string({
      char: "p",
      description: "Optional JSON-formatted data payload to submit with the test",
      required: false,
    }),
    contentType: Flags.string({
      char: "c",
      description: "Optional content-type for the test payload",
      required: false,
    }),
  };

  async run() {
    const {
      args: { flow },
      flags: { tail, payload, contentType },
    } = await this.parse(TestCommand);

    const result = await gqlRequest({
      document: gql`
        mutation testIntegrationFlow(
          $id: ID!
          $payload: String
          $contentType: String
        ) {
          testIntegrationFlow(
            input: { id: $id, payload: $payload, contentType: $contentType }
          ) {
            testIntegrationFlowResult {
              flow {
                id
              }
              execution {
                id
              }
            }
            errors {
              field
              messages
            }
          }
        }
      `,
      variables: {
        id: flow,
        payload,
        contentType,
      },
    });
    const executionId = result.testIntegrationFlow.testIntegrationFlowResult.execution.id;
    console.log(`Execution ID: ${executionId}`);
    if (tail) {
      await this.tailLogs(executionId);
    }
  }

  private async tailLogs(executionId: string) {
    const { flags } = await this.parse(TestCommand);

    let nextCursor: string | undefined = undefined;
    while (true) {
      await ux.wait(500);

      const result: any = await this.fetchLogs(executionId, nextCursor);
      if (result === undefined) continue;

      const { logs, cursor, executionComplete } = result;

      nextCursor = cursor;

      ux.table(
        logs,
        {
          timestamp: {},
          severity: {
            minWidth: 12,
          },
          message: {},
        },
        { ...flags, "no-header": true },
      );

      if (executionComplete) return;
    }
  }

  private async fetchLogs(
    executionId: string,
    nextCursor?: string,
  ): Promise<FetchLogsResult | undefined> {
    const results = await gqlRequest({
      document: gql`
        query listIntegrationTestLogs($executionId: ID!, $nextCursor: String) {
          logs(
            executionResult: $executionId
            after: $nextCursor
            orderBy: { field: TIMESTAMP, direction: ASC }
          ) {
            edges {
              node {
                timestamp
                severity
                message
              }
              cursor
            }
          }
        }
      `,
      variables: {
        executionId,
        nextCursor,
      },
    });

    const { edges }: { edges: { node: LogNode; cursor?: string }[] } = results.logs;
    if (!edges || edges.length === 0) {
      return undefined;
    }

    const logs = edges.map(({ node }) => node);
    const executionComplete = logs.reduce(
      (result: boolean, { message }) => result || message.startsWith("Ending Instance"),
      false,
    );

    const { cursor } = edges[edges.length - 1];
    return { logs, cursor, executionComplete };
  }
}
