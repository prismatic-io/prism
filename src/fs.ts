import { promises as fs } from "fs";

export const exists = async (path: string) => {
  return fs.access(path).then(
    () => true,
    () => false,
  );
};

export const readStdin = async () => {
  return new Promise((resolve, reject) => {
    process.stdin
      .on("readable", () => {
        const result = process.stdin.read();
        if (result !== null) {
          process.stdin.removeAllListeners();
          resolve(result.toString());
        }
      })
      .on("error", reject);
  });
};

export { fs };
