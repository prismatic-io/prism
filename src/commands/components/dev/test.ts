import { promisify } from "util";
import dotenv from "dotenv";
import { Command, Flags, CliUx } from "@oclif/core";
import inquirer, { DistinctQuestion } from "inquirer";
import { snakeCase, upperCase, kebabCase } from "lodash";
import { serverTypes } from "@prismatic-io/spectral"; // FIXME: Get rid of this and stop exporting it in Spectral.
import {
  loadEntrypoint,
  createComponentPackage,
  publishDefinition,
  uploadConnectionIcons,
  uploadFile,
  validateDefinition,
  checkPackageSignature,
} from "../../../utils/component/publish";
import { displayLogs } from "../../../utils/execution/logs";
import {
  buildComponentTestHarnessIntegration,
  componentTestIntegrationName,
  ComponentTestInfo,
} from "../../../utils/integration/definition";
import { importDefinition } from "../../../utils/integration/import";
import {
  deleteIntegration,
  runIntegrationFlow,
} from "../../../utils/integration/invoke";
import { pollForActiveConfigVarState } from "../../../utils/integration/query";
import { Expression } from "../../../utils/integration/export";
import { exists } from "../../../fs";
import { whoAmI } from "../../../utils/user/query";
import { spawnProcess } from "../../../utils/process";
import {
  printFinalStepResults,
  writeFinalStepResults,
} from "../../../utils/execution/stepResults";
import { deleteComponentByKey } from "../../../utils/component/deleteByKey";

const setTimeoutPromise = promisify(setTimeout);

const envVarCase = (name: string): string =>
  upperCase(snakeCase(name)).replace(/\s+/g, "_");

const toInquirerInputType = (
  type: string,
  collection: string | undefined
): Required<DistinctQuestion["type"]> => {
  if (collection) {
    // FIXME: Improve prompting instead of instantly bailing to editor.
    return "editor";
  }

  switch (type) {
    case "boolean":
      return "checkbox";
    case "password":
      return "password";
    case "code":
      return "editor";
    default:
      return "input";
  }
};

const getInputQuestion = ({
  key,
  label,
  type,
  collection,
  default: defaultValue,
}: serverTypes.Input): DistinctQuestion => ({
  type: toInquirerInputType(type, collection),
  name: key,
  message: `${label}:`,
  when: (answers) => {
    const envVar = envVarCase(key);
    const exists = envVar in process.env;
    if (exists) {
      const value = process.env[envVar] ?? "";
      answers[key] = collection
        ? { type: "complex", value: JSON.parse(value) }
        : { type: "value", value };
    }
    return !exists;
  },
  filter: (value) => {
    if (type === "connection") {
      return { type: "configVar", value };
    }

    if (collection) {
      return {
        type: "complex",
        value: JSON.parse(value),
      };
    }

    return {
      type: "value",
      value,
    };
  },
  default: () => (type === "connection" ? "testConnection" : defaultValue),
});

interface PromptAnswers {
  action: serverTypes.Action;
  actionInputs: Record<string, Expression>;
  connection?: serverTypes.Connection;
  connectionInputs?: Record<string, Expression>;
}

const valuesFromAnswers = ({
  action,
  actionInputs,
  connection,
  connectionInputs = {},
}: PromptAnswers): Pick<ComponentTestInfo, "actionInfo" | "connectionInfo"> => {
  const actionValues = Object.entries(actionInputs).reduce<
    ComponentTestInfo["actionInfo"]["values"]
  >(
    (result, [key, value]) => ({
      ...result,
      [key]: value,
    }),
    {}
  );

  const connectionValues = Object.entries(connectionInputs).reduce<
    Required<ComponentTestInfo>["connectionInfo"]["values"]
  >(
    (result, [key, value]) => ({
      ...result,
      [key]: value,
    }),
    {}
  );

  const connectionInfo = connection
    ? { key: connection.key, values: connectionValues }
    : undefined;

  return {
    actionInfo: { key: action.key, values: actionValues },
    connectionInfo,
  };
};

export default class TestCommand extends Command {
  static description =
    "Run an action of a component within a test integration in the integration runner";
  static flags = {
    envPath: Flags.string({
      required: false,
      default: ".env",
      char: "e",
      description: "Path to dotenv file to load for supplying testing values",
    }),
    build: Flags.boolean({
      required: false,
      default: true,
      allowNo: true,
      char: "b",
      description: "Build the component prior to testing",
    }),
    "output-file": Flags.string({
      required: false,
      char: "o",
      description: "Output the results of the action to a specified file",
    }),
    "print-results": Flags.boolean({
      required: false,
      default: false,
      description: "Print the results of the action to stdout",
    }),
    "clean-up": Flags.boolean({
      required: false,
      default: false,
      description:
        "Clean up the integration and temporary component after running the action",
    }),
  };

  async run() {
    const {
      flags: {
        envPath,
        build,
        "output-file": outputFile,
        "print-results": printResults,
        "clean-up": cleanUp,
      },
    } = await this.parse(TestCommand);

    // Save the current working directory, so we can return later after moving to dist/
    const cwd = process.cwd();

    if (build) {
      console.log("Building component...");
      await spawnProcess(["npm", "run", "build"], {});
    }

    if (await exists(envPath)) {
      const { error } = dotenv.config({ path: envPath, override: true });
      if (error) {
        CliUx.ux.error(`Failed to load specified dotenv file: ${error}`, {
          exit: 1,
        });
      }
    }

    CliUx.ux.action.start("Validating Component");

    const { name } = await whoAmI();
    if (!name) {
      CliUx.ux.error(
        "Failed to determine the name of the currently logged in user.",
        {
          exit: 1,
        }
      );
    }

    const testingKey = kebabCase(name);

    const definition = await loadEntrypoint();
    const { key: componentKey, public: isPublic } = definition;
    definition.key = `${componentKey}-${testingKey}-testing`;
    definition.display.label = `${definition.display.label} ${name} Testing`;
    await validateDefinition(definition);

    const packagePath = await createComponentPackage();
    const signatureMatches = await checkPackageSignature(
      definition,
      packagePath
    );

    CliUx.ux.action.stop();

    if (!signatureMatches) {
      CliUx.ux.action.start("Publishing Component");

      const { iconUploadUrl, packageUploadUrl, connectionIconUploadUrls } =
        await publishDefinition(definition);

      const {
        display: { iconPath },
      } = definition;
      await uploadFile(iconPath, iconUploadUrl);
      await uploadConnectionIcons(definition, connectionIconUploadUrls);
      await uploadFile(packagePath, packageUploadUrl);

      CliUx.ux.action.stop();
    }

    const publishedTimestamp = Date.now();

    const actions = definition.actions || {};

    const { action } = await inquirer.prompt<{ action: serverTypes.Action }>({
      type: "list",
      name: "action",
      message: "Action:",
      choices: Object.entries(actions).map(
        ([
          key,
          {
            display: { label },
          },
        ]) => ({
          name: label,
          value: key,
          short: key,
        })
      ),
      default: Object.keys(actions)[0],
      filter: (value: string) => actions[value],
    });

    const { inputs } = action;

    // Ask for values of action's inputs
    const actionInputs = await inquirer.prompt(
      inputs.map((i) => getInputQuestion(i))
    );

    const answers = {
      action,
      actionInputs,
    };

    // Ask about Connection to test if there is a connection type input
    const hasConnection = inputs.some(({ type }) => type === "connection");
    if (hasConnection) {
      const connections = definition.connections || [];
      const { connection } = await inquirer.prompt<{
        connection: serverTypes.Connection;
      }>({
        type: "list",
        name: "connection",
        message: "Connection:",
        choices: connections.map(({ key, label }) => ({
          name: label,
          value: key,
          short: key,
        })),
        default: Object.keys(connections)[0],
        filter: (value) => {
          const [connection] = connections.filter(({ key }) => value === key);
          return connection;
        },
      });

      // Prompt for connection's inputs
      const { inputs } = connection;
      const connectionInputs = await inquirer.prompt(
        inputs
          .filter(({ shown }) => shown === undefined || shown === true)
          .map((i) => getInputQuestion(i))
      );

      Object.assign(answers, { connection, connectionInputs });
    }

    const { actionInfo, connectionInfo } = valuesFromAnswers(answers);

    CliUx.ux.action.start("Assembling test integration");

    // FIXME: Wait for version to be available but due to issues we have to do a static wait.
    const wait = 5000 - (Date.now() - publishedTimestamp);
    if (wait > 0) {
      await setTimeoutPromise(wait);
    }

    // Build up YAML
    const harnessYaml = await buildComponentTestHarnessIntegration({
      integrationInfo: {
        name: componentTestIntegrationName(componentKey, name),
      },
      componentInfo: { key: definition.key, isPublic: isPublic ?? false },
      actionInfo,
      connectionInfo,
    });

    CliUx.ux.action.stop();
    CliUx.ux.action.start("Updating test Integration");

    const {
      integrationId,
      flows: [{ id: flowId }],
      pendingAuthorizations,
    } = await importDefinition(harnessYaml);

    CliUx.ux.action.stop();

    // Prompt for user authorization of pending connections
    if (pendingAuthorizations.length > 0) {
      const [{ id, url }] = pendingAuthorizations;
      if (!url) {
        throw new Error(
          "Did not receive a valid URL for authorization. Verify your Connection inputs."
        );
      }

      CliUx.ux.url("Authorize URL", url);
      await CliUx.ux.anykey(
        "Press any key to open your browser and authorize the Connection"
      );
      await CliUx.ux.open(url);

      CliUx.ux.action.start("Waiting for Connection authorization");

      await pollForActiveConfigVarState(integrationId, id);

      CliUx.ux.action.stop();
    }

    CliUx.ux.action.start("Running test Integration");

    const { executionId } = await runIntegrationFlow({ integrationId, flowId });

    await displayLogs(executionId);

    if (outputFile) {
      process.chdir(cwd);
      console.log(`Writing step results to ${outputFile}`);
      await writeFinalStepResults(executionId, outputFile);
    }

    if (printResults) {
      await printFinalStepResults(executionId);
    }

    if (cleanUp) {
      CliUx.ux.action.start(`Cleaning up test Integration (${integrationId})`);
      await deleteIntegration(integrationId);
      CliUx.ux.action.stop();

      CliUx.ux.action.start(`Cleaning up test component (${definition.key})`);
      await deleteComponentByKey(definition.key);
      CliUx.ux.action.stop();
    }

    CliUx.ux.action.stop();
  }
}
