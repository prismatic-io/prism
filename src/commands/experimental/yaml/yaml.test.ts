import { describe, it, expect } from "bun:test";
import GenerateIntegrationFromYAMLCommand from "./index.js";
import fs from "fs";
import path from "path";
import { readFile } from "fs-extra";
import { walkDir } from "../../../fs.js";

const CONVERT_GENERATION_TIMEOUT = 5 * 60 * 5; // 5 minutes

describe("YAML CNI generation tests", () => {
  const basePath = process.env.PWD ?? process.cwd();
  const commandPath = path.resolve("src/commands/experimental/yaml");
  const tempPath = path.resolve(`${commandPath}/temp`);

  if (!fs.existsSync(tempPath)) {
    fs.mkdirSync(tempPath);
  }

  describe("test convert integration", () => {
    it(
      "should generate successfully",
      async () => {
        process.chdir(tempPath);

        await GenerateIntegrationFromYAMLCommand.run([
          "../fixtures/specs/test-integration.yaml",
          "--folder=testIntegration",
        ]);

        expect(process.cwd()).toStrictEqual(tempPath);

        process.chdir(basePath);
      },
      CONVERT_GENERATION_TIMEOUT,
    );

    it("should match scaffolding snapshots", async () => {
      process.chdir(tempPath);
      const targets = await walkDir("testIntegration", [
        ".npmrc",
        ".png",
        "webpack.config.js",
        "package.json",
        "src/componentRegistry.ts",
      ]);
      for (const target of targets) {
        const contents = await readFile(target, "utf-8");
        expect(contents).toMatchSnapshot(target);
      }

      process.chdir(basePath);
    });
  });
});
