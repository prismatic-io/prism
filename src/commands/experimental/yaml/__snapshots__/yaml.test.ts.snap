// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`YAML CNI generation tests test-integration should match scaffolding snapshots: test-integration/jest.config.js 1`] = `
"module.exports = {
  preset: "ts-jest",
  testEnvironment: "node",
};
"
`;

exports[`YAML CNI generation tests test-integration should match scaffolding snapshots: test-integration/.npmrc 1`] = `
"@component-manifests:registry=https://app.dev.prismatic-dev.io/packages/npm
"
`;

exports[`YAML CNI generation tests test-integration should match scaffolding snapshots: test-integration/.prettierignore 1`] = `
"# Ignore artifacts:
.spectral
build
coverage
"
`;

exports[`YAML CNI generation tests test-integration should match scaffolding snapshots: test-integration/.prettierrc 1`] = `
"{}
"
`;

exports[`YAML CNI generation tests test-integration should match scaffolding snapshots: test-integration/tsconfig.json 1`] = `
"{
  "compilerOptions": {
    "target": "es6",
    "lib": ["esnext"],
    "strict": true,
    "esModuleInterop": true,
    "forceConsistentCasingInFileNames": true,
    "module": "esnext",
    "moduleResolution": "node"
  },
  "include": ["src", ".spectral/*"]
}
"
`;

exports[`YAML CNI generation tests test-integration should match scaffolding snapshots: test-integration/.spectral/index.ts 1`] = `
"import type { ComponentManifest, ConfigPage, ScopedConfigVar } from "@prismatic-io/spectral";

// @ts-ignore
import type { configPages, componentRegistry, userLevelConfigPages, scopedConfigVars } from "../src";

type IsAny<T> = 0 extends 1 & T ? true : false;

type TConfigPages =
  IsAny<typeof configPages> extends true
    ? { [key: string]: ConfigPage }
    : typeof configPages;

type TUserLevelConfigPages =
  IsAny<typeof userLevelConfigPages> extends true
    ? { [key: string]: ConfigPage }
    : typeof userLevelConfigPages;

type TComponentRegistry =
  IsAny<typeof componentRegistry> extends true
    ? { [key: string]: ComponentManifest }
    : typeof componentRegistry;

type TScopedConfigVarMap =
  IsAny<typeof scopedConfigVars> extends true
    ? { [key: string]: ScopedConfigVar }
    : typeof scopedConfigVars;

declare module "@prismatic-io/spectral" {
  // eslint-disable-next-line @typescript-eslint/no-empty-interface
  interface IntegrationDefinitionConfigPages extends TConfigPages {}

  // eslint-disable-next-line @typescript-eslint/no-empty-interface
  interface IntegrationDefinitionUserLevelConfigPages
    extends TUserLevelConfigPages {}

  // eslint-disable-next-line @typescript-eslint/no-empty-interface
  interface IntegrationDefinitionComponentRegistry extends TComponentRegistry {}

  // eslint-disable-next-line @typescript-eslint/no-empty-interface
  interface IntegrationDefinitionScopedConfigVars extends TScopedConfigVarMap {}
}

"
`;

exports[`YAML CNI generation tests test-integration should match scaffolding snapshots: test-integration/src/configPages.ts 1`] = `
"import { configPage, configVar, connectionConfigVar, dataSourceConfigVar } from "@prismatic-io/spectral";

export const configPages = {
  "Connections Page": configPage({
    tagline: "",
    elements: {
      "Slack Connection": connectionConfigVar({
        stableKey: "Slack Connection",
        dataType: "connection",
        connection: {
          component: "slack",
          key: "oauth2",
          values: {
            "authorizeUrl": {
              value: "https://slack.com/oauth/v2/authorize",
              permissionAndVisibilityType: "embedded",
              visibleToOrgDeployer: false,
            },
            "clientId": {
              value: "7610303900993.7610321662833",
              permissionAndVisibilityType: "embedded",
              visibleToOrgDeployer: false,
            },
            "clientSecret": {
              value: "b6acb317302a8ce8a01fbf24ab371338",
              permissionAndVisibilityType: "embedded",
              visibleToOrgDeployer: false,
            },
            "isUser": {
              value: false,
              permissionAndVisibilityType: "embedded",
              visibleToOrgDeployer: false,
            },
            "revokeUrl": {
              value: "https://slack.com/api/auth.revoke",
              permissionAndVisibilityType: "embedded",
              visibleToOrgDeployer: false,
            },
            "scopes": {
              value: "chat:write chat:write.public chat:write.customize channels:read groups:read im:read mpim:read users:read",
              permissionAndVisibilityType: "embedded",
              visibleToOrgDeployer: false,
            },
            "signingSecret": {
              value: "32cf662b1f05eb9a4472ef6900f0bb53",
              permissionAndVisibilityType: "embedded",
              visibleToOrgDeployer: false,
            },
            "tokenUrl": {
              value: "https://slack.com/api/oauth.v2.access",
              permissionAndVisibilityType: "embedded",
              visibleToOrgDeployer: false,
            },
          },
        },
      }),
    },
  }),
  "Config Vars Page": configPage({
    tagline: "",
    elements: {
      "Simple string": configVar({
        stableKey: "Simple string",
        dataType: "string",
        description: "undefined",
        permissionAndVisibilityType: "organization",
        visibleToOrgDeployer: true,
        defaultValue: "The quick brown fox jumps over the lazy dog.",
      }),
      "Simple number": configVar({
        stableKey: "Simple number",
        dataType: "number",
        description: "undefined",
        permissionAndVisibilityType: "embedded",
        defaultValue: 1,
      }),
      "Simple list": configVar({
        stableKey: "Simple list",
        dataType: "string",
        description: "undefined",
        permissionAndVisibilityType: "embedded",
        collectionType: "valuelist", defaultValue: ["abc", "def", "ghi"],
      }),
      "Simple key value list": configVar({
        stableKey: "Simple key value list",
        dataType: "string",
        description: "undefined",
        permissionAndVisibilityType: "embedded",
        collectionType: "keyvaluelist", defaultValue: [{ "key": "key1", "value": "myvalue" }, { "key": "key2", "value": "myothervalue" }, { "key": "key3", "value": "myotherothervalue" }],
      }),
    },
  }),
  "Data Sources Page": configPage({
    tagline: "",
    elements: {
      "Slack Channels Data Source": dataSourceConfigVar({
        stableKey: "Slack Channels Data Source",
        dataType: "picklist",
        dataSource: {
          component: "slack",
          key: "selectChannels",
          values: {
            "connection": {
              configVar: "Slack Connection",
              permissionAndVisibilityType: "embedded",
              visibleToOrgDeployer: false,
            },
            "includeImChannels": {
              value: false,
              permissionAndVisibilityType: "embedded",
              visibleToOrgDeployer: false,
            },
            "includeMultiPartyImchannels": {
              value: false,
              permissionAndVisibilityType: "embedded",
              visibleToOrgDeployer: false,
            },
            "includePrivateChannels": {
              value: false,
              permissionAndVisibilityType: "embedded",
              visibleToOrgDeployer: false,
            },
            "includePublicChannels": {
              value: true,
              permissionAndVisibilityType: "embedded",
              visibleToOrgDeployer: false,
            },
            "showIdInDropdown": {
              value: false,
              permissionAndVisibilityType: "embedded",
              visibleToOrgDeployer: false,
            },
          },
        },
      }),
    },
  }),
}
  ;
"
`;

exports[`YAML CNI generation tests test-integration should match scaffolding snapshots: test-integration/src/componentRegistry.ts 1`] = `
"import { componentManifests } from "@prismatic-io/spectral";
import branch from "@component-manifests/branch";
import code from "@component-manifests/code";
import log from "@component-manifests/log";
import loop from "@component-manifests/loop";
import scheduleTriggers from "@component-manifests/schedule-triggers";
import slack from "@component-manifests/slack";
import webhookTriggers from "@component-manifests/webhook-triggers";

export const componentRegistry = componentManifests({
  branch,
  code,
  log,
  loop,
  scheduleTriggers,
  slack,
  webhookTriggers,
})
  ;
"
`;

exports[`YAML CNI generation tests test-integration should match scaffolding snapshots: test-integration/src/index.ts 1`] = `
"import { integration } from "@prismatic-io/spectral";
import flows from "./flows";
import { configPages } from "./configPages";
import { componentRegistry } from "./componentRegistry";

export { configPages } from "./configPages";
export { componentRegistry } from "./componentRegistry";

export default integration({
  name: "Test Integration",
  description: "",
  iconPath: "icon.png",
  flows,
  configPages,
  componentRegistry,
});
"
`;

exports[`YAML CNI generation tests test-integration should match scaffolding snapshots: test-integration/src/flows/basic.ts 1`] = `
"import { flow } from "@prismatic-io/spectral";

export const basic = flow({
  name: "Basic",
  stableKey: "basic",
  description: "",
  isSynchronous: false,
  endpointSecurityType: "customer_optional",
  onTrigger: {
    component: "webhookTriggers",
    key: "webhook",
    values: {
      body: { value: "", },
      contentType: { value: "", },
      headers: { value: [], },
      statusCode: { value: "", },
    },
  },
  onExecution: async (context, params) => {
    const { configVars } = context;
    const postMessage = await context.components.slack.postMessage({
      channelName: "general",
      connection: configVars["Slack Connection"],
      debug: false,
      message: "Hello this is my Slack message",
      messageId: "",
      username: "myTestSlackBot",
    });
    const writeLogMessage = await context.components.log.writeLog({
      level: "info",
      message: "Here is a simple log message at the end of my flow",
    });
    return { data: writeLogMessage };
  },
})
  ;

export default basic;
"
`;

exports[`YAML CNI generation tests test-integration should match scaffolding snapshots: test-integration/src/flows/index.ts 1`] = `
"import { basic } from "./basic";
import { stepConfigVarReferences } from "./stepConfigVarReferences";
import { loopBranchTests } from "./loopBranchTests";

export default [basic, stepConfigVarReferences, loopBranchTests]
  ;
"
`;

exports[`YAML CNI generation tests test-integration should match scaffolding snapshots: test-integration/src/flows/stepConfigVarReferences.ts 1`] = `
"import { flow } from "@prismatic-io/spectral";

export const stepConfigVarReferences = flow({
  name: "Step & configVar references",
  stableKey: "step-config-var-references",
  description: "",
  isSynchronous: false,
  endpointSecurityType: "customer_optional",
  onTrigger: {
    component: "webhookTriggers",
    key: "webhook",
    values: {
      body: { value: "", },
      contentType: { value: "", },
      headers: { value: [], },
      statusCode: { value: "", },
    },
  },
  onExecution: async (context, params) => {
    const { configVars } = context;
    const codeBlock = await context.components.code.runCode({
      code: \`module.exports = async ({ logger, configVars }, stepResults) => {
  return { data: { testKey: "My return value" } };
};
\`,
    });
    const templateReferences = await context.components.log.writeLog({
      level: "info",
      message: \`The response of the code block action was: \${codeBlock.data.testKey}. This config var value is: \${configVars["Simple string"]}.\`,
    });
    const directConfigVarReference = await context.components.log.writeLog({
      level: "info",
      message: configVars["Simple string"],
    });
    const directStepReference = await context.components.log.writeLog({
      level: "info",
      message: codeBlock.data.testKey,
    });
    const dataSourceReference = await context.components.log.writeLog({
      level: "info",
      message: configVars["Slack Channels Data Source"],
    });
    return { data: dataSourceReference };
  },
})
  ;

export default stepConfigVarReferences;
"
`;

exports[`YAML CNI generation tests test-integration should match scaffolding snapshots: test-integration/src/flows/loopBranchTests.ts 1`] = `
"import { isEqual } from "@prismatic-io/spectral/dist/conditionalLogic";
import { flow } from "@prismatic-io/spectral";

export const loopBranchTests = flow({
  name: "Loop & branch tests",
  stableKey: "loop-branch-tests",
  description: "",
  isSynchronous: false,
  endpointSecurityType: "customer_optional",
  schedule: {
    value: "*/10 * * * *",
  },
  onExecution: async (context, params) => {
    const { configVars } = context;
    const listUsers = await context.components.slack.listUsers({
      connection: configVars["Slack Connection"],
      cursor: "",
      debug: false,
      limit: "",
      teamId: "",
    });
    const loopOverItems: { data: unknown[] } = { data: [] };
    for (const loopOverItemsItem of listUsers.data.members) {
      const writeLogMessage = await context.components.log.writeLog({
        level: "info",
        message: loopOverItemsItem.name,
      });

      /* This string is the name of the branch that this conditional block resolves to. */
      let branchOnExpression: string;
      if (isEqual(loopOverItemsItem.name, "slackbot")) {
        const codeBlock = await context.components.code.runCode({
          code: \`module.exports = async ({ logger, configVars }, stepResults) => {
  return { data: true };
};
\`,
        });
        branchOnExpression = "it is slackbot";
      } else {
        break;
      }
      loopOverItems.data.push(branchOnExpression);
    }

    const loopNTimes: { data: unknown[] } = { data: [] };
    for (let loopNTimesIdx = 0; loopNTimesIdx < 100; loopNTimesIdx++) {
      const writeLogMessage2 = await context.components.log.writeLog({
        level: "info",
        message: loopOverItems.data[0],
      });

      const codeBlock2 = await context.components.code.runCode({
        code: \`module.exports = async ({ logger, configVars }, stepResults) => {
  return { data: 100 };
};
\`,
      });

      loopNTimes.data.push(codeBlock2.data);
    }

    const writeLogMessage3 = await context.components.log.writeLog({
      level: "info",
      message: loopNTimes.data[5],
    });
    const selectExecutedStepResult = await context.components.branch.selectExecutedStepResult({ stepResults: ["aTestValue",], });
    return { data: selectExecutedStepResult };
  },
})
  ;

export default loopBranchTests;
"
`;
